src/Workspaces/Core/Portable/FindSymbols/FindReferences/Finders/PropertySymbolReferenceFinder.cs

```diff
@@ -147,79 +142,83 @@ private static bool IsForEachProperty(IPropertySymbol symbol)
 
-            var syntaxRoot = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
-            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
-            var semanticFacts = document.GetLanguageService<ISemanticFactsService>();
+            SyntaxNode indexerReference;
+            var matched = false;
+            var reason = CandidateReason.None;
 
-            // Now that we have Doc Comments in place, We are searching for References in the Trivia as well by setting descendIntoTrivia: true
-            var indexerMemberCrefs = syntaxRoot.DescendantNodes(descendIntoTrivia: true)
-                                               .Where(syntaxFacts.IsIndexerMemberCRef);
+            if (syntaxFacts.IsElementAccessExpression(node))
+            {
+                // For an ElementAccessExpression the indexer we are looking for is the argumentList component.
+                syntaxFacts.GetPartsOfElementAccessExpression(node, out var expression, out indexerReference);
 
-            var locations = ArrayBuilder<FinderLocation>.GetInstance();
+                if (expression != null && symbolsMatch(expression, semanticModel).matched)
+                {
+                    // Element access with explicit member name (allowed in VB).
+                    // We have already added a reference location for the member name identifier, so skip this one.
+                    return (indexerReference, matched, reason);
+                }
 
-            foreach (var node in indexerMemberCrefs)
+                (matched, reason) = symbolsMatch(node, semanticModel);
+            }
+            else if (syntaxFacts.IsConditionalAccessExpression(node))
             {
-                cancellationToken.ThrowIfCancellationRequested();
+                // For a ConditionalAccessExpression the whenNotNull component is the indexer reference we are looking for
+                syntaxFacts.GetPartsOfConditionalAccessExpression(node, out var expression, out indexerReference);
 
-                var (matched, reason) = symbolsMatch(node, semanticModel);
-                if (matched)
+                if (syntaxFacts.IsInvocationExpression(indexerReference) && symbolsMatch(indexerReference, semanticModel).matched)
                 {
-                    var location = node.SyntaxTree.GetLocation(new TextSpan(node.SpanStart, 0));
-                    var symbolUsageInfo = GetSymbolUsageInfo(node, semanticModel, syntaxFacts, semanticFacts, cancellationToken);
-                    locations.Add(new FinderLocation(
-                        node, new ReferenceLocation(document, null, location, isImplicit: false, symbolUsageInfo, GetAdditionalFindUsagesProperties(node, semanticModel, syntaxFacts), candidateReason: reason)));
+                    // Element access with explicit member name (allowed in VB).
+                    // We have already added a reference location for the member name identifier, so skip this one.
```

Comment:
you're calling `symbolsMatch(indexerReference, semanticModel)` both in the if-condition and after the `if` executes, which is pretty darn confusing.  It needs to be run in all paths.  Can you instead run it up front. If it doesn't match, we just `return default;` immediately.   Then you do the rest of the work.  This also applies to the check on line 188